src/common/src/mlib/str.h

```diff
@@ -565,4 +642,488 @@ mstr_contains_any_of(mstr_view str, mstr_view needle)
+   if (new_len < str->len) {
+      // We didn't resize first, so resize now. We are shrinking the string, so this
+      // will never fail, and does not create any uninitialized memory:
+      mlib_check(mstr_resize_for_overwrite(str, new_len));
+      mut = str->data + splice_pos;
+   }
+   // Insert the new data
+   memcpy(mut, insert.data, insert.len);
+   return true;
+}
+
+/**
+ * @brief Append a string to the end of some other string.
+ *
+ * @param str The string to be modified
+ * @param suffix The suffix string to be appended onto `*str`
+ * @return true If the operation was successful
+ * @return false Otherwise
+ *
+ * If case of failure, `*str` is not modified.
+ */
+static inline bool
+mstr_append(mstr *str, mstr_view suffix)
+{
+   return mstr_splice(str, str->len, 0, suffix);
+}
+
+#define mstr_append(Into, Suffix) mstr_append((Into), mstr_view_from((Suffix)))
+
+/**
+ * @brief Append a single character to the given string object
+ *
+ * @param str The string object to be updated
+ * @param c The single character that will be inserted at the end
+ * @return true If the operation succeeded
+ * @return false Otherwise
+ *
+ * In case of failure, the string is not modified.
+ */
+static inline bool
+mstr_pushchar(mstr *str, char c)
+{
+   mstr_view one = mstr_view_data(&c, 1);
+   return mstr_append(str, one);
+}
+
+/**
+ * @brief Replace every occurrence of `needle` in `str` with `sub`
+ *
+ * @param str The string object to be updated
+ * @param needle The non-empty needle string to be searched for.s
+ * @param sub The string to be inserted in place of each `needle`
+ * @return true If the operation succeeds
+ * @return false Otherwise
+ *
+ * @warning If the `needle` string is empty, then the program will terminate!
+ * @note If the operation fails, the content of `str` is an unspecified but valid
+ * string.
+ */
+static inline bool
+mstr_replace(mstr *str, mstr_view needle, mstr_view sub)
+{
+   mlib_check(needle.len, neq, 0, because, "Trying to replace an empty string will result in an infinite loop");
+   // Scan forward, starting from the first position:
+   size_t off = 0;
+   while (1) {
+      // Find the next occurrence, starting from the scan offset
+      off = mstr_find(*str, needle, off);
+      if (off == SIZE_MAX) {
+         // No more occurrences.
+         return true;
+      }
+      // Replace the needle string with the new value
+      if (!mstr_splice(str, off, needle.len, sub)) {
+         return false;
+      }
+      // Advance over the length of the replacement string, so we don't try to
+      // infinitely replace content if the replacement itself contains the needle
+      // string
+      if (mlib_add(&off, sub.len)) {
+         // Integer overflow while advancing the offset. No good.
+         return false;
+      }
+   }
+}
+
+/**
+ * @brief Like `mstr_sprintf_append`, but accepts the va_list directly.
+ */
+MLIB_IF_GNU_LIKE(__attribute__((format(printf, 2, 0))))
```

Comment:
This (and other occurrances of the) printf attribute in this file may need to special-case for GCC mingw compatibility by using `gnu_printf`, e.g. see "third issue" PR description of https://github.com/mongodb/libmongocrypt/pull/1054#event-19180404991 and (note: the version check is likely unnecessary at this point given other occurrances of this pattern): https://github.com/mongodb/mongo-c-driver/blob/d9c26f49e75d3de746a690db9c81ff5b4f6e21b0/src/libbson/src/bson/macros.h#L306-L312